pal/src/httpapi_sl.c

```diff
@@ -0,0 +1,492 @@
+        }
+        if (apiH->x509Certificate != NULL) {
+            free(apiH->x509Certificate);
+        }
+        if (apiH->x509PrivateKey != NULL) {
+            free(apiH->x509PrivateKey);
+        }
+        HTTPClient_destroy(apiH->cli);
+    }
+
+    if (apiH != NULL) {
+        free(apiH);
+    }
+}
+
+HTTPAPI_RESULT HTTPAPI_ExecuteRequest(HTTP_HANDLE handle,
+        HTTPAPI_REQUEST_TYPE requestType, const char* relativePath,
+        HTTP_HEADERS_HANDLE httpHeadersHandle, const unsigned char* content,
+        size_t contentLength, unsigned int* statusCode,
+        HTTP_HEADERS_HANDLE responseHeadersHandle,
+        BUFFER_HANDLE responseContent)
+{
+    HTTPAPI_Object * apiH = (HTTPAPI_Object *)handle;
+    HTTPClient_Handle cli = apiH->cli;
+    int i;
+    int ret;
+    int offset;
+    size_t cnt;
+    char contentBuf[CONTENT_BUF_LEN] = {0};
+    uint32_t contentBufLen = CONTENT_BUF_LEN;
+    char *hname;
+    char *hvalue;
+    const char *method;
+    bool moreFlag;
+    HTTPClient_extSecParams esParams = {NULL, NULL, SL_SSL_CA_CERT};
+
+    method = getHttpMethod(requestType);
+
+    if ((cli == NULL) || (method == NULL) || (relativePath == NULL)
+            || (statusCode == NULL) || (responseHeadersHandle == NULL)) {
+        LogError("Invalid arguments: handle=%p, requestType=%d, "
+            "relativePath=%p, statusCode=%p, responseHeadersHandle=%p",
+            handle, (int)requestType, relativePath, statusCode,
+            responseHeadersHandle);
+        return (HTTPAPI_INVALID_ARG);
+    }
+    else if (HTTPHeaders_GetHeaderCount(httpHeadersHandle, &cnt)
+            != HTTP_HEADERS_OK) {
+        LogError("Cannot get header count");
+        return (HTTPAPI_QUERY_HEADERS_FAILED);
+    }
+
+    /*
+     * We need to call HTTPClient_connect here as opposed to
+     * HTTPAPI_CreateConnection because we might have set some TLS
+     * options.
+     */
+    if (apiH->isConnected == false) {
+        esParams.clientCert = apiH->x509Certificate;
+        esParams.privateKey = apiH->x509PrivateKey;
+        ret = HTTPClient_connect(apiH->cli, apiH->prefixedHostName,
+                &esParams, 0);
+        if (ret < 0) {
+            LogError("HTTPClient_connect failed, ret=%d", ret);
+            return (HTTPAPI_OPEN_REQUEST_FAILED);
+        }
+        else {
+            apiH->isConnected = true;
+        }
+    }
+
+    /* Send the request headers */
+    while (cnt--) {
+        ret = HTTPHeaders_GetHeader(httpHeadersHandle, cnt, &hname);
+        if (ret != HTTP_HEADERS_OK) {
+            LogError("Cannot get request header %d", cnt);
+            return (HTTPAPI_QUERY_HEADERS_FAILED);
+        }
+
+        ret = splitHeader(hname, &hvalue);
+
+        if (ret == 0) {
+            /*
+             * HOST and Content-Length headers are set by HTTPClient
+             * automatically. Note that Content-Length = 0 never gets sent.
+             */
+            if ((stringcasecmp(hname, "content-length") != 0) &&
```

Comment:
>stringcasecmp [](start = 17, length = 13)

I'm not sure why they used their own `stringcasecmp()` instead of the standard and widely-available `strcasecmp()`.